include/flux/op/cartesian_product.hpp

```diff
@@ -216,6 +255,37 @@ struct sequence_traits<detail::cartesian_product_adaptor<Bases...>>
         }(std::index_sequence_for<Bases...>{});
     }
 
+    template <std::size_t I, typename Self, typename Function,
+              typename... PartialCursor>
+    static constexpr auto for_each_while_impl(Self& self,
+                                              Function&& func,
+                                              PartialCursor&&... partial_cursor)
+        -> std::tuple<bool, detail::cartesian_product_partial_cursor_t<Self, I>>
+    {
+        // We need to iterate right to left.
+        if constexpr (I == sizeof...(Bases) - 1) {
+            bool keep_going = true;
+            auto this_current = flux::for_each_while(std::get<I>(self.bases_),
+                [&](auto&& elem) {
+                    keep_going = std::invoke(func,
+                        cursor_t<Self>(FLUX_FWD(partial_cursor)..., FLUX_FWD(elem)));
```

Comment:
This took me so long to figure out what was going on, because it seems obviously wrong but the tests still pass. But that's because the only cartesian product test for this uses `flux::ints` where this just happens to be the right thing.

The issues here are:
* `partial_cursor...` are not actual partial cursors, they're elements
* we need to pass an element (i.e. a tuple of elements) to `func`, not a cursor.

So if you try to do something like this, it doesn't compile:
```cpp
flux::cartesian_product(flux::ints(0, 3), flux::from_range(std::vector{10, 20, 30}))
    .for_each([](auto&&){});
```

Have to do this instead:

```suggestion
                        element_t<Self>(FLUX_FWD(partial_cursor)..., FLUX_FWD(elem)));
```

And then rename `partial_cursor` to `elems` or something.

Reply:
Thanks - Tristan pointed this out early. I added tests for this in a240dcc8d6aa4a1023e28ee94d83d3d68d640d82, and fixed it similar to how you described in 02e73b3c0d97a414f600dfe5391832cf12a67809.

Reply:
There's no test for short-circuiting - if you just `return true;` in all the relevant places, all the tests pass.

Reply:
Adding a test for short circuiting seems wise.

Reply:
Test added in 155a5ddaa8a6d0780d5b09ffe134bb694b990fc.